File: bridgeAddon/addon/globalPlugins/nvdaMcpBridge/domain/ports/clock.py
# nvdaMcpBridge domain -- the Clock port.
# Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.

from __future__ import annotations

from abc import ABC, abstractmethod


class Clock(ABC):
	"""Monotonic time + sleep, injected wherever time is read or waited on.

	Fakes advance only when told, so timeout/heartbeat/inactivity behaviour is
	exercised without ever sleeping in real time.
	"""

	@abstractmethod
	def monotonic(self) -> float:
		"""Seconds from an arbitrary fixed origin; only differences matter."""

	@abstractmethod
	def sleep(self, seconds: float) -> None:
		"""Block for ``seconds`` (a fake may make this an instant clock advance)."""

File: server/domain/ports/clock.go
// screenreader-mcp domain -- the Clock port.
// Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.
//
// ROLE: domain port, monotonic time plus sleeping.
// IMPLEMENTED BY: adapters/system_clock.go, and fakes/clock.go in tests.
// USED BY: the bridge client's request deadlines and the connection controller's heartbeat.
package ports

import "time"

type Clock interface {
	// Now is monotonic; only differences between readings are meaningful.
	Now() time.Time

	Sleep(d time.Duration)
}
